src/mlpack/tests/loss_functions_test.cpp

```diff
@@ -748,4 +749,136 @@ BOOST_AUTO_TEST_CASE(MarginRankingLossTest)
+
+  // Test the Backward function.
+  module.Backward(input, target, output);
+  BOOST_REQUIRE_CLOSE(arma::as_scalar(arma::accu(output)), 0.505909, 1e-3);
+  BOOST_REQUIRE_EQUAL(output.n_rows, input.n_rows);
+  BOOST_REQUIRE_EQUAL(output.n_cols, input.n_cols);
+  CheckMatrices(output, expectedOutput, 0.1);
+}
+
+/**
+ * Simple test for the MultiLabel Softmargin Loss function.
+ */
+BOOST_AUTO_TEST_CASE(MultiLabelSoftMarginLossMeanReductionTest)
+{
+  arma::mat input, target, output, expectedOutput;
+  double loss;
+  MultiLabelSoftMarginLoss<> module(false);
+
+  input = arma::mat("0.1778 0.0957 0.1397 0.1203 0.2403 0.1925 -0.2264 -0.3400 "
+      "-0.3336");
+  target = arma::mat("0 1 0 1 0 0 0 0 1");
+  // Calculated using torch.nn.MultiLabelSoftMarginLoss(reduction='mean').
+  expectedOutput = arma::mat("0.0605 -0.0529 0.0594 -0.0522 0.0622 0.0609 "
+      "0.0493 0.0462 -0.0647");
+  input.reshape(3, 3);
+  target.reshape(3, 3);
+  expectedOutput.reshape(3, 3);
+
+  // Test the Forward function. Loss should be 0.716095.
+  // Value calculated using torch.nn.MultiLabelSoftMarginLoss(reduction='mean').
+  loss = module.Forward(input, target);
+  BOOST_REQUIRE_CLOSE(loss, 0.716095, 1e-3);
+
+  // Test the Backward function.
+  module.Backward(input, target, output);
+  BOOST_REQUIRE_CLOSE(arma::as_scalar(arma::accu(output)), 0.168636, 1e-3);
+  BOOST_REQUIRE_EQUAL(output.n_rows, input.n_rows);
+  BOOST_REQUIRE_EQUAL(output.n_cols, input.n_cols);
+  CheckMatrices(output, expectedOutput, 0.1);
+}
+
+/**
+ * Simple test for the MultiLabel Softmargin Loss function.
+ */
+BOOST_AUTO_TEST_CASE(MultiLabelSoftMarginLossWeightedSumReductionTest)
+{
+  arma::mat input, target, output, expectedOutput;
+  arma::rowvec weights;
+  double loss;
+  weights = arma::mat("1 2 3");
+  MultiLabelSoftMarginLoss<> module(true, weights);
+
+  input = arma::mat("0.1778 0.0957 0.1397 0.2256 0.1203 0.2403 0.1925 0.3144 "
+      "-0.2264 -0.3400 -0.3336 -0.8695");
+  target = arma::mat("0 1 0 1 1 0 0 0 0 0 1 0");
+  // Calculated using torch.nn.MultiLabelSoftMarginLoss(reduction='sum').
+  expectedOutput = arma::mat("0.1814 -0.1587 0.1783 -0.1479 -0.3133 0.3732 "
+      "0.3653 0.3853 0.4436 0.4158 -0.5826 0.2954");
+  input.reshape(4, 3);
+  target.reshape(4, 3);
+  expectedOutput.reshape(4, 3);
+
+  // Test the Forward function. Loss should be 5.35057.
+  // Value calculated using torch.nn.MultiLabelSoftMarginLoss(reduction='sum').
+  loss = module.Forward(input, target);
+  BOOST_REQUIRE_CLOSE(loss, 5.35057, 1e-3);
+
+  // Test the Backward function.
+  module.Backward(input, target, output);
+  BOOST_REQUIRE_CLOSE(arma::as_scalar(arma::accu(output)), 1.43577, 1e-3);
+  BOOST_REQUIRE_EQUAL(output.n_rows, input.n_rows);
+  BOOST_REQUIRE_EQUAL(output.n_cols, input.n_cols);
+  CheckMatrices(output, expectedOutput, 0.1);
+}
+
+/**
+ * Simple test for the MultiLabel Softmargin Loss function.
+ */
+BOOST_AUTO_TEST_CASE(MultiLabelSoftMarginLossWeightedMeanReductionTest)
```

Comment:
I think we should have only 2 test cases i.e. MultiLabelSoftMarginLossWeightedTest and MultiLabelSoftMarginLossTest containing test for both reductions. The same was done for other layers in ann.

Reply:
I can do this. Just one question. I did this because some of the other loss functions have 2 separate tests - one with reduction and another without, thus I thought it would make more sense to split the tests into 4 parts - for all combinations of weights and reductions. Don't you think that would make it easier going forward with the weighted losses for other loss functions? I am okay with whatever you suggest though.

Reply:
I think only KL Divergence does that. Other than that, most have a single loss function. You can refer to this [one](https://github.com/mlpack/mlpack/blob/944f15048d73178ad98de409f6ae7e3a4c329d82/src/mlpack/tests/loss_functions_test.cpp#L624) for example. Also most of the open PRs follow the same. 

> one with mean reduction and another without, thus I thought it would make more sense to split the tests into 4 parts - for all combinations of weights and reductions.

In Ann layer tests, we generally add tests in one test case unless a new parameter was added. So Reduction is a common parameter so we can 1 test case testing both sum and mean. At least that way we won't end up 4 times the test cases. Let me know if this makes sense.

> Don't you think that would make it easier going forward with the weighted losses for other loss functions?

I think having a different test for weighted makes sense because that way when other loss functions also support weighted we won't have to change original tests but only add new ones.

Reply:
Okay! Cool. I will re-arrange the code for the test cases into 2 blocks.

Reply:
Awesome! Thanks again. :)